diffrax/integrate.py

```diff
@@ -195,6 +203,168 @@ def _apply_history(x):
             error_order,
             state.controller_state,
         )
+
+        # Finding all of the potential discontinuity roots
+        if delays is not None:
+
+            def _maybe_find_discontinuity(sub_tprev, sub_tnext):
+                dense_ts = jnp.stack([tprev, tnext])
+                dense_infos = jtu.tree_map(lambda x: x[None], dense_info)
+                dense_discont = DenseInterpolation(
+                    ts=dense_ts,
+                    ts_size=2,
+                    interpolation_cls=solver.interpolation_cls,
+                    infos=dense_infos,
+                    direction=1,
+                )
+                flat_delays = jtu.tree_leaves(delays.delays)
+                _gs = []
+                for delay in flat_delays:
+                    g = (
+                        lambda t, delay=delay: t
+                        - delay(t, dense_discont.evaluate(t), args)
+                        - state.discontinuities
+                    )
+                    _gs.append(g)
+
+                def _find_discontinuity():
+                    # Start by doing a cheap bisection search to reduce
+                    # over the stored-discontinuity dimension.
+
+                    def _cond_fun(_val):
+                        _, _, _pred, _, _ = _val
+                        return _pred
+
+                    def _body_fun(_val):
+                        _ta, _tb, _, _step, max_step = _val
+                        _step = _step + 1
+                        _tmid = _ta + 0.5 * (_tb - _ta)
+                        _gas = jnp.stack([jnp.sign(g(_ta)) for g in _gs])
+                        _gmids = jnp.stack([jnp.sign(g(_tmid)) for g in _gs])
+                        _gbs = jnp.stack([jnp.sign(g(_tb)) for g in _gs])
+                        _any_left = jnp.any(_gas != _gmids)
+                        _next_ta = jnp.where(_any_left, _ta, _tmid)
+                        _next_tb = jnp.where(_any_left, _tmid, _tb)
+                        _pred = (
+                            jnp.any(jnp.sum(_gas != _gbs, axis=1) > 1) | _step
+                            > max_step
+                        )
+                        return _next_ta, _next_tb, _pred, _step, max_step
+
+                    _init_val = (sub_tprev, sub_tnext, True, 0, 400)
```

Comment:
What if we hit `max_steps` here? I think we need a mechanism in this whole discontinuity procedure to allow the discontinuity-finding to fail. (Namely, reject the step and use some more naive way of picking `tnext`, e.g. `tprev` plus 0.1 times the current interval length or something.)

Reply:
If we hit `max_steps` that would just means that the NewtonSolver would take the lead with the current root approximation ? Speaking from experience, since each `sub_tprev` and `sub_tnext` used are pretty small intervals this would probably not happen ?